### PR TITLE
Windows: add script to build a .msi package

### PR DESCRIPTION
**Pull request by laszlocsomor**
Also add a test, though there's no build rule to
wrap it yet.

The generated installer:
- is a standard .msi installer that users are
  familiar with
- installs Bazel at a short and safe path (no
  space in path)
- updates the PATH envvar
- handles uninstall, upgrading, parallel installs:
  backwards-compatible versions can be upgraded
  (but not downgraded), while incompatible
  versions may be installed side by side
  (see `guids.txt` comments)

Example use:

```
c:\src\bazel>powershell scripts\packages\wix\make_msi.ps1 ..\bazel-releases\0.26.1\bazel-0.26.1-windows-x86_64.exe
(...)
INFO[make_msi_lib.ps1 11:41:33] Creating wixobj
INFO[make_msi_lib.ps1 11:41:33] Creating msi
INFO[make_msi_lib.ps1 11:41:42] Done: ..\bazel-releases\0.26.1\bazel-0.26.1-windows-x86_64.msi

c:\src\bazel>start ..\bazel-releases\0.26.1\bazel-0.26.1-windows-x86_64.msi
```

TODO(@laszlocsomor) before we can use this:
- figure out if we should bundle vc_redist with
  the Bazel installer
- figure out the UpgradeCode for such a combined
  bundle

### CODE REVIEW

**Comment by laszlocsomor** [comment]
# Installer

bazel-setup-1
bazel-setup-2
bazel-setup-3

# Add or remove programs

bazel-setup-4

# PATH envvar updated automatically (also on uninstall)

bazel-setup-5

File: scripts/packages/wix/make_msi_lib.ps1 [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    if ("$ReleaseName" -match "rc[0-9]+$") {
+        return "$ReleaseName".Substring(0, "$ReleaseName".LastIndexOf("rc"))
+    } else {
+        return "$ReleaseName"
+    }
+}
+
+# Downloads and extracts the WiX Toolkit.
+# Returns the path where the tools are (e.g. "candle.exe").
+function Download-Wix {
```
Comment (meteorcloudy) [inline scripts/packages/wix/make_msi_lib.ps1]: Why do we need to use Bazel to download Wix? I think we can just use powershell commands to download an extract WiX Toolkit?

Comment (laszlocsomor) [reply scripts/packages/wix/make_msi_lib.ps1]: Yes, `start-bitstranser` and `expand-archive` are the relevant commands.
Alas, I get permission errors when trying to use them.

Comment (meteorcloudy) [reply scripts/packages/wix/make_msi_lib.ps1]: For downloading, we can also use 
`(New-Object Net.WebClient).DownloadFile` 
See https://github.com/bazelbuild/continuous-integration/blob/master/buildkite/setup-windows.ps1#L140

Comment (laszlocsomor) [reply scripts/packages/wix/make_msi_lib.ps1]: Ah, here's the reason: https://powershell.org/forums/topic/bits-transfer-with-github/
Let me see if I can rewrite it with System.Net.WebClient.

Comment (meteorcloudy) [reply scripts/packages/wix/make_msi_lib.ps1]: Is the permission error thrown by `start-bitstranser` or `expand-archive`?

Comment (meteorcloudy) [reply scripts/packages/wix/make_msi_lib.ps1]: Sounds good!

Comment (laszlocsomor) [reply scripts/packages/wix/make_msi_lib.ps1]: Done. Thanks for the suggestion!

**Comment by meteorcloudy** [comment]
Awesome! Maybe this installer could help us installing extra dependencies of Bazel. For example, any dependencies of the future Windows sandbox.
FYI @rongjiecomputer

File: scripts/packages/wix/make_msi_lib.ps1 [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
```
Comment (aiuto) [inline scripts/packages/wix/make_msi_lib.ps1]: I'm not really comfortable with downloading Wix here.
This is fine for now, but I think we should vendor in tools like this, especially when we get to code signing.
And that brings up the whole question of if we should continue to bundle all our dependencies in the distribution tarball (no) or call them out by reference (yes).

**Comment by aiuto** [comment]
You left a very thorny TODO
- figure out if we should bundle vc_redist with the Bazel installer
- figure out the UpgradeCode for such a combined bundle

I think the answer is no. Do not bundle it IF you want to a .MSI installer built with Wix. When you try to get two things in the MSI it leads to odd rollback cases if both do not cleanly install/update/delete together. An alternative to bundling is to say we require vc_redist and let MSI decide where to get it.\

Another thing to consider is if you want to write the script to run Wix in powershell or Python. We might find it easier to maintain the script in Python, since only a few people know powershell.

You put this in  scripts/packages/wix, but is wix really the important thing in the path? I think scripts/packages/msi would be the clearer meaning.

File: scripts/packages/wix/make_msi_lib.ps1 (hunk 2 above)
Comment (laszlocsomor) [reply scripts/packages/wix/make_msi_lib.ps1]: Thanks for the observation.

I think downloading is simpler than vendoring, and unlike vendoring it raises no questions about licencing. So I'd like to keep it like so for now.

**Comment by laszlocsomor** [comment]
Thanks for your insights @aiuto !

Re: bundling vc\_redist: I don't think we need to, for now. Most Bazel users will probably install VC++ which installs these DLLs. Arguably not everyone will, sure, but if people complain about this requirement then we can revisit this question.

Re: UpgradeCode: if we don't bundle vc\_redist, then this is moot.

Re: package path: good idea, I'll rename the "wix" directory to "msi".

Re: python: I prefer to stay with PowerShell. I agree more Bazel core developers know Python than PS, but:
- PS is a better tool for this job (calling binaries, generating GUIDs is just easier)
- PS is native on Windows (we'd be fine with a Bash script if this were a Linux installer), and
- PS is simple and well-documented, so maintenance shouldn't involve a steep learning curve. Writing this script took about 1.5 days, which involved learning PS.

**Comment by laszlocsomor** [comment]
@aiuto : PTAL

**Comment by laszlocsomor** [comment]
oh wait, you already approved, thanks :)

@meteorcloudy , does this LGTY?

**Pull request closed without merging**